numpy/core/numeric.py

```diff
@@ -1097,10 +1113,24 @@ def tensordot(a, b, axes=2):
     newshape_b = (N2, int(multiply.reduce([bs[ax] for ax in notin])))
     oldb = [bs[axis] for axis in notin]
 
+    if out is not None:
+        correct = True
+        if isinstance(out, np.ndarray):
+            if out.shape != tuple(olda+oldb):
+                correct = False
+            else:
+                out = out.reshape((newshape_a[0], newshape_b[1]))
+        else: 
+            correct = False
```

Comment:
This should be a TypeError, I think

Reply:
Does it actually have to be a C order array? I think we can just keep those error details to the `np.dot` function.

Reply:
I think that checking only shape should be enough. `np.dot` will check other details.

Reply:
The danger is that `reshape` may create a copy, so `F` contiguous arrays may be accepted by `dot`, but not necessarily work here (at least not without a copy back into `out`). In any case, just make sure we have tests for these things, then it is easy to try around to see how many checks are needed.